### PR TITLE
feat: add `clDice` loss

### LINKED ISSUE #5938: feature request : please add cldice loss
**Is your feature request related to a problem? Please describe.**
[clDice - a Novel Topology-Preserving Loss Function for Tubular Structure Segmentation](https://openaccess.thecvf.com/content/CVPR2021/papers/Shit_clDice_-_A_Novel_Topology-Preserving_Loss_Function_for_Tubular_Structure_CVPR_2021_paper.pdf)
The abstract of this paper:⬇️
> Accurate segmentation of tubular, network-like structures, such as vessels, neurons, or roads, is relevant to many fields of research. For such structures, the topology is their most important characteristic; particularly preserving connectedness: in the case of vascular networks, missing a connected vessel entirely alters the blood-flow dynamics. We introduce a novel similarity measure termed centerlineDice (short clDice), which is calculated on the intersection of the segmentation masks and their (morphological) skeleta. We theoretically prove that clDice guarantees topology preservation up to homotopy equivalence for binary 2D and 3D segmentation. Extending this, we propose a computationally efficient, differentiable loss function (soft-clDice) for training arbitrary neural segmentation networks. We benchmark the soft-clDice loss on five public datasets, including vessels, roads and neurons (2D and 3D). Training on soft-clDice leads to segmentation with more accurate connectivity information, higher graph similarity, and better volumetric scores.

**Describe the solution you'd like**
[Official code repo](https://github.com/jocpae/clDice)
Please integrate the code in Monai project, that will help lots of researchers who working on vessels, neurons, tree or tabular structure. This official code repo has already implemented 2D and 3D version clDice loss, it will be easy to merge to current MONAI project. Thank you.

### PR DESCRIPTION
**Pull request by SauravMaheshkar**
Fixes #5938

### Description

This PR aims to add the `SoftclDiceLoss` and the `SoftDiceclDiceLoss` from [clDice - a Novel Topology-Preserving Loss Function for Tubular Structure Segmentation](https://openaccess.thecvf.com/content/CVPR2021/papers/Shit_clDice_-_A_Novel_Topology-Preserving_Loss_Function_for_Tubular_Structure_CVPR_2021_paper.pdf)

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [x] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/losses/cldice.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+            torch.sum(skel_pred[:, 1:, ...]) + self.smooth
+        )
+        tsens = (torch.sum(torch.multiply(skel_true, y_pred)[:, 1:, ...]) + self.smooth) / (
+            torch.sum(skel_true[:, 1:, ...]) + self.smooth
+        )
+        cl_dice: torch.Tensor = 1.0 - 2.0 * (tprec * tsens) / (tprec + tsens)
+        return cl_dice
+
+
+class SoftDiceclDiceLoss(_Loss):
```
Comment (wyli) [inline monai/losses/cldice.py:139]: thanks @SauravMaheshkar! please add some unit tests like https://github.com/Project-MONAI/MONAI/blob/dev/tests/test_giou_loss.py

Comment (SauravMaheshkar) [reply monai/losses/cldice.py:149]: Addressed in 35f8c287647993f4a46d12cac1c4a7f2088382a5

**Review by wyli — APPROVED** [review]
Thanks, it looks good to me.

File: tests/test_cldice_loss.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+import numpy as np
+import torch
+from parameterized import parameterized
+
+from monai.losses import SoftclDiceLoss, SoftDiceclDiceLoss
+
+TEST_CASES = [
+    [  # shape: (1, 4), (1, 4)
+        {"y_pred": torch.ones((100, 3, 256, 256)), "y_true": torch.ones((100, 3, 256, 256))},
```
Comment (wyli) [inline tests/test_cldice_loss.py:21]: hi @SauravMaheshkar could you please help make these test cases smaller? the github test runner doesn't seem to have enough memory for these https://github.com/Project-MONAI/MONAI/actions/runs/5655050610/job/15327390836

Comment (SauravMaheshkar) [reply tests/test_cldice_loss.py:21]: Okay will try

Comment (SauravMaheshkar) [reply tests/test_cldice_loss.py:21]: If that fixes the OOM error then it LGTM 😄
